File: quiver-parser.js
const fs = require('fs').promises
const path = require('path')

// const baseDir = '../lib'

if (process.argv.length !== 3) {
  console.error(`Usage: node ${path.basename(process.argv[1])} <base-dir>`)
  process.exit(1)
}

const baseDir = process.argv[2]

const notebooks = [
  {
    name: 'Inbox',
    uuid: 'Inbox',
  },
]

const convertToISOString = ts => new Date(ts * 1000).toISOString()

const readAndParseJson = async (filename) => {
  const content = await fs.readFile(path.join(baseDir, filename), 'utf-8')
  return JSON.parse(content)
}

const getNote = async (notebookName, notebookUuid, noteDirName) => {
  const path = `${notebookUuid}.qvnotebook/${noteDirName}`

  const meta = await readAndParseJson(`${path}/meta.json`)
  const content = await readAndParseJson(`${path}/content.json`)

  return {
    uuid: meta.uuid,
    title: meta.title,
    contents: content.cells.map((cell) => cell.data).join('\\n\\n'),
    notebook: notebookName,
    created_at: convertToISOString(meta.created_at),
    updated_at: convertToISOString(meta.updated_at),
  }
}

const parseNotes = async (notebookName, notebookUuid) => {
  const allFiles = await fs.readdir(
    path.join(baseDir, `${notebookUuid}.qvnotebook`)
  )
  const noteDirs = allFiles.filter((filename) => filename.endsWith('.qvnote'))

  const notes = await Promise.all(
    noteDirs.map((noteDir) => getNote(notebookName, notebookUuid, noteDir))
  )

  return notes
}

const parseNotebooks = async (parentName, children) => {
  for await ({ uuid: notebookUuid, children: childrenNotebook } of children) {
    const { name } = await readAndParseJson(
      `${notebookUuid}.qvnotebook/meta.json`
    )

    const nameWithParent = [parentName, name].join(' > ')
    notebooks.push({
      name: nameWithParent,
      uuid: notebookUuid,
    })

    if (childrenNotebook) {
      await parseNotebooks(nameWithParent, childrenNotebook)
    }
  }
}

const main = async () => {
  const root = await readAndParseJson('meta.json')
  await parseNotebooks('', root.children)

  const notes = (
    await Promise.all(
      notebooks.map(({ name, uuid }) => {
        return parseNotes(name, uuid)
      })
    )
  ).flat()

  console.log(JSON.stringify(notes, null, 2))
}

main()

File: quiver-converter/quiver-parser.py
import os
import sys
import json
from datetime import datetime, timezone

if len(sys.argv) != 2:
  name = sys.argv[0]
  print(f"Usage: python3 {name} <base-dir>")
  sys.exit(1)

base_dir = sys.argv[1]

notebooks = [
  {
    "name": "Inbox",
    "uuid": "Inbox",
  },
]


def convert_to_iso8601(ts):
  return datetime.fromtimestamp(ts, timezone.utc).isoformat()


def read_and_parse_json(filename):
  with open(os.path.join(base_dir, filename), "r", encoding="utf-8") as file:
    return json.load(file)


def get_note(notebook_name, notebook_uuid, note_dir_name):
  path = f"{notebook_uuid}.qvnotebook/{note_dir_name}"

  meta = read_and_parse_json(f"{path}/meta.json")
  content = read_and_parse_json(f"{path}/content.json")

  return {
    "uuid": meta["uuid"],
    "title": meta["title"],
    "contents": "\\n\\n".join(cell["data"] for cell in content["cells"]),
    "notebook": notebook_name,
    "created_at": convert_to_iso8601(meta["created_at"]),
    "updated_at": convert_to_iso8601(meta["updated_at"]),
  }


def parse_notes(notebook_name, notebook_uuid):
  all_files = os.listdir(os.path.join(base_dir, f"{notebook_uuid}.qvnotebook"))
  note_dirs = [filename for filename in all_files if filename.endswith(".qvnote")]
  notes = [get_note(notebook_name, notebook_uuid, note_dir) for note_dir in note_dirs]
  return notes


def parse_notebooks(parent_name, children):
  for child in children:
    notebook_uuid = child["uuid"]

    child_metadata = read_and_parse_json(f"{notebook_uuid}.qvnotebook/meta.json")

    name_with_parent = f"{parent_name} > {child_metadata['name']}"

    notebooks.append(
      {
        "name": name_with_parent,
        "uuid": notebook_uuid,
      }
    )

    children_notebook = child.get("children")

    if children_notebook:
      parse_notebooks(name_with_parent, children_notebook)


if __name__ == "__main__":
  root = read_and_parse_json("meta.json")
  parse_notebooks("", root["children"])

  notes = []
  for notebook in notebooks:
    notes.extend(parse_notes(notebook["name"], notebook["uuid"]))

  print(json.dumps(notes, indent=2, ensure_ascii=False))
